File: C++/2-Medium/00133._克隆图/00133._克隆图.cpp

#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Node{
    public:
    int val;
    vector<Node*> neighbors;

    Node(){
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val){
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors){
        val = _val;
        neighbors = _neighbors;
    }
};

class Solution {
    public:
    unordered_map<Node*, Node*> clonedNodes;

    Node* cloneGraph(Node* node){
        if(node == nullptr){
            return nullptr;
        }

        if(clonedNodes.find(node) != clonedNodes.end()){
            return clonedNodes[node];
        }

        Node* clone = new Node(node->val);
        clonedNodes[node] = clone;
        for(Node* neighbor : node->neighbors){
            clone->neighbors.push_back(cloneGraph(neighbor));
        }

        return clone;
    }
};

int main() {
    // 示例：创建一个简单的图 (1 -- 2 -- 3 -- 4 -- 1)
    Node* node1 = new Node(1);
    Node* node2 = new Node(2);
    Node* node3 = new Node(3);
    Node* node4 = new Node(4);
    node1->neighbors = {node2, node4};
    node2->neighbors = {node1, node3};
    node3->neighbors = {node2, node4};
    node4->neighbors = {node1, node3};

    Solution solution;
    Node* clonedGraph = solution.cloneGraph(node1);

    // 输出克隆图的验证信息（此处简化，实际需要遍历图来验证）
    cout << "Cloned graph's node 1 value: " << clonedGraph->val << endl;

    return 0;
}

File: Python/2-Medium/00133._克隆图/00133._克隆图.py
from typing import Dict, List, Optional

# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors: Optional[List['Node']] = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        def clone(node: 'Node', copies: Dict[int, 'Node']) -> 'Node':
            """
            使用 DFS 递归克隆图中的每个节点及其边

            参数:
            - node: 当前正在访问的节点
            - copies: 存储已克隆节点的哈希表，键为原节点的值，值为克隆的节点实例
            
            返回:
            - 克隆的节点实例
            """
            # 如果节点为空，返回 None
            if not node:
                return None
            
            # 如果节点已经被克隆过，直接从哈希表中返回克隆的节点
            if node.val in copies:
                return copies[node.val]
            
            # 创建当前节点的克隆
            clone_node = Node(node.val)
            copies[node.val] = clone_node
            
            # 递归克隆当前节点的所有邻居，并添加到克隆节点的邻居列表中
            for neighbor in node.neighbors:
                clone_node.neighbors.append(clone(neighbor, copies))
            
            return clone_node
        
        # 使用哈希表存储所有克隆的节点，以便快速访问和防止无限递归
        copies = {}
        return clone(node, copies)

# 示例代码用于测试
# 创建图的示例节点
node1 = Node(1)
node2 = Node(2)
node3 = Node(3)
node4 = Node(4)

# 建立邻居关系
node1.neighbors = [node2, node4]
node2.neighbors = [node1, node3]
node3.neighbors = [node2, node4]
node4.neighbors = [node1, node3]

# 克隆图
solution = Solution()
cloned_graph = solution.cloneGraph(node1)

# 检查克隆的结果（由于环境限制，此处不进行实际打印，仅示意代码运行逻辑）
# 实际使用时，可以通过遍历 cloned_graph 并打印节点及其邻居的值来验证克隆结果
